lib/cached_network_image.dart

```diff
@@ -86,81 +90,177 @@ class CacheManager {
 
   CacheManager._();
 
-  SharedPreferences prefs;
-  Map<String, CacheObject> cacheData;
+  SharedPreferences _prefs;
+  Map<String, CacheObject> _cacheData;
+  DateTime lastCacheClean;
+
   static Object _lock = new Object();
 
   ///Shared preferences is used to keep track of the information about the files
   _init() async {
-    prefs = await SharedPreferences.getInstance();
+    _prefs = await SharedPreferences.getInstance();
 
     //get saved cache data from shared prefs
-    var jsonCacheString = prefs.getString("lib_cached_image_data");
-    cacheData = new Map();
+    var jsonCacheString = _prefs.getString("lib_cached_image_data");
+    _cacheData = new Map();
     if (jsonCacheString != null) {
       Map jsonCache = JSON.decode(jsonCacheString);
       jsonCache.forEach((key, data) {
-        cacheData[key] = new CacheObject.fromMap(data);
+        _cacheData[key] = new CacheObject.fromMap(key, data);
       });
     }
+
+    // Get data about when the last clean action has been performed
+    var cleanMillis = _prefs.getInt("lib_cached_image_data_last_clean");
+    if (cleanMillis != null) {
+      lastCacheClean = new DateTime.fromMillisecondsSinceEpoch(cleanMillis);
+    } else {
+      lastCacheClean = new DateTime.now();
+      _prefs.setInt("lib_cached_image_data_last_clean",
+          lastCacheClean.millisecondsSinceEpoch);
+    }
   }
 
+  bool _isStoringData = false;
+  bool _shouldStoreDataAgain = false;
+  Object _storeLock = new Object();
   ///Store all data to shared preferences
   _save() async {
+
+    if(!(await _canSave())){
+      return;
+    }
+
+    await _cleanCache();
+    await _saveDataInPrefs();
+  }
+
+  Future<bool> _canSave() async {
+    return await synchronized(_storeLock, (){
+      if(_isStoringData){
+        _shouldStoreDataAgain = true;
+        return false;
+      }
+      _isStoringData = true;
+      return true;
+    });
+  }
+
+  Future<bool> _shouldSaveAgain() async{
+    return await synchronized(_storeLock, (){
+      if(_shouldStoreDataAgain){
+        _shouldStoreDataAgain = false;
+        return true;
+      }
+      _isStoringData = false;
+      return false;
+    });
+  }
+
+  _saveDataInPrefs() async{
     Map json = new Map();
+
     await synchronized(_lock, () {
-      cacheData.forEach((key, cache) {
+      _cacheData.forEach((key, cache) {
         json[key] = cache._map;
       });
     });
-    prefs.setString("lib_cached_image_data", JSON.encode(json));
+    _prefs.setString("lib_cached_image_data", JSON.encode(json));
+
+    if(await _shouldSaveAgain()){
+      await _saveDataInPrefs();
```

Comment:
looks like `_saveDataInPrefs` calls `_saveDataInPrefs` again and uses number of special flags to control that behavior and avoid infinite loop.
what is the reason we need to do `Save` and then again `Save`?

Reply:
The logic I use here is the following.
Imagine you download 5 files simultaneously. You would end up in 'Save' 5 times. If the first one is still busy saving the other 4 just indicate that something has changed and return. The first one will save one more time in the end, resulting in 2 saves instead of 5. It also only cleans the cache once.